sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuTransitionOverrides.scala

```diff
@@ -527,7 +527,8 @@ class GpuTransitionOverrides extends Rule[SparkPlan] {
    */
   private def insertColumnarToGpu(plan: SparkPlan): SparkPlan = {
     val nonQueryStagePlan = GpuTransitionOverrides.getNonQueryStagePlan(plan)
-    if (nonQueryStagePlan.supportsColumnar && !nonQueryStagePlan.isInstanceOf[GpuExec]) {
+    if (nonQueryStagePlan.supportsColumnar && !nonQueryStagePlan.isInstanceOf[GpuExec] &&
+      !nonQueryStagePlan.isInstanceOf[AdaptiveSparkPlanExec]) {
```

Comment:
Never insert a `HostColumnarToGpu` for `AdaptiveSparkPlanExec`. Does that work on all of the versions of Spark that we support?

Reply:
I didn't think AdaptiveSparkPlanExec would appear in front of a scan but instead in front of querystage or subquery node where AQE can examine the shuffle partitions and decide how many tasks to run in the subsequent stage.  Is that not the case?

Otherwise I'm not sure how we can know for sure whether an AdaptiveSparkPlanExec is really producing GPU output.  If the AdaptiveSparkPlan is not fully planned yet then I don't know if we can guarantee we know it will be.  I suppose one other approach is to use a plan tag approach where the shuffle exec marks the underlying AdaptiveSparkPlanExec node when it converts it to columnar to note that we should not inject a host columnar transition on that.

Reply:
There are two comments that I have here and I think you are responding to both here so I will try to do the same.

My first concern is that with this change we are essentially saying that all AdaptiveSparkPlanExec nodes that `supportsColumnar` output will produce GPU output. I think that is okay because I think we are the only ones that will set AdaptiveSparkPlanExec to be columnar. I just would like a comment or something explaining what is happening because it feel brittle.

My second concern is more of a performance issue compared to anything. We equate `supportsColumnar` with this is running on the GPU. I think that is a fairly good assumption most of the time, but a scan will output columnar data that is not on the GPU.  This code runs before columnar to row transitions are inserted into the plan. Which means if I do a join on a parquet file input, the first stage of the plan will look like a parquet read -> shuffle.  I don't know how AQE will plan things out for the child plan, but if it ever looks like read -> AdaptiveSparkPlanExec -> shuffle we could end up doing a shuffle on the GPU when the read might not be on the GPU.  But I think that is rare enough it is fine. Again it is more of having a comment here might be nice to explain what is happening any why.

Reply:
> I just would like a comment or something explaining what is happening because it feel brittle.

I'll add comments around this.  I agree it feels brittle, but I wasn't sure there was a better way to do this.  If Spark starts doing significant columnar processing in nodes we cannot replace (like AdaptiveSparkPlanExec), we've got a lot of work to do to keep things straight. :sweat_smile: 

> if it ever looks like read -> AdaptiveSparkPlanExec -> shuffle we could end up doing a shuffle on the GPU when the read might not be on the GPU.

Yeah this is related to the comment above.  We don't really have a good way to distinguish whether we're CPU or GPU for columnar execs that we cannot replace with GPU execs.

Reply:
I suppose for the AdaptiveSparkPlanExec is already columnar, we could do a similar technique to https://github.com/NVIDIA/spark-rapids/pull/9134/files#diff-389f8b5c364533aaab72c65af450b2a2a9c7c944112527455ba34af06a188c72R664-R665 where we try to see if the head node looks like it's going to be on the GPU in the adaptive sub-plan.  I think that would resolve the concern about non-GPU columnar read in that scenario.

Reply:
I am fine with just comments for now.

Reply:
I ended up tackling this with a new utility method in GpuOverrides that tries to predict whether an AdaptiveSparkPlanExec appears it will be translated to the GPU based on its current subplan.  Also refactored the code to use some more descriptive names which IMHO makes the code more readable without needing more comments.  If you feel more comments are still warranted, happy to add them.